torchao/quantization/prototype/qat/utils.py

```diff
@@ -143,3 +179,61 @@ def _choose_qparams_per_token_asymmetric(
     zero_point = torch.clamp(zero_point, qmin, qmax).round()
 
     return scale.to(scales_precision), zero_point.to(zero_points_precision)
+
+def _forward_pre_hook_handler(
+    mod: torch.nn.Linear,
+    prehook: Callable,
+    handler: torch.utils.hooks.RemovableHandle,
+):
+    """
+    Store a 2-tuple (prehook function, handler) as an attribute on the given linear module.
+    """
+    setattr(mod, _QAT_LINEAR_SUBCLASS_INPUT_PREHOOK, (prehook, handler))
+
+def _unwrap_affine_fake_quantized_tensor(t: torch.Tensor):
+    """
+    Return the original, non-fake-quantized float tensor from a `AffineFakeQuantizedTensor`.
+    """
+    # avoid circular dependencies
+    from torchao.quantization.prototype.qat.affine_fake_quantized_tensor import (
+        AffineFakeQuantizedTensor,
+    )
+    assert isinstance(t, AffineFakeQuantizedTensor)
+    return t.original_tensor
+
+def _is_linear_with_fq_weight(mod: torch.nn.Module, *args):
+    """
+    Return whether this is a nn.Linear module with `AffineFakeQuantizeTensor` weights.
+    """
+    # avoid circular dependencies
+    from torchao.quantization.prototype.qat.affine_fake_quantized_tensor import (
+        AffineFakeQuantizedTensor,
+    )
+    if not isinstance(mod, torch.nn.Linear) or not hasattr(mod, "weight"):
+        return False
+    weight = mod.weight
+    return isinstance(weight, AffineFakeQuantizedTensor)
+
+def _enable_fake_quant(mod: torch.nn.Module, enable: bool):
+    """
+    Enable or disable fake quantization in the activations and weights of a `nn.Linear` module.
+    """
+    from torchao.quantization.prototype.qat.affine_fake_quantized_tensor import (
+        AffineFakeQuantizedTensor,
+    )
+    if not _is_linear_with_fq_weight(mod):
+        return
+    weight = mod.weight
+    assert isinstance(weight, AffineFakeQuantizedTensor)
+    weight.fake_quant_enabled = enable
+
+    # Enable/disable input fake quant
+    if hasattr(mod, _QAT_LINEAR_SUBCLASS_INPUT_PREHOOK):
+        (prehook, handle) = getattr(mod, _QAT_LINEAR_SUBCLASS_INPUT_PREHOOK)
+        if enable:
+            if handle is None:
+                handle = mod.register_forward_pre_hook(prehook)
+                _forward_pre_hook_handler(mod, prehook, handle)
```

Comment:
I assume this is because we are passing `_forward_pre_hook_handler` to `_get_linear_subclass_inserter`, maybe we can clean this up if we just create a helper function just for QAT

Reply:
done